tokio-stream/src/wrappers/mpsc_bounded.rs

```diff
@@ -67,6 +67,27 @@ impl<T> Stream for ReceiverStream<T> {
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         self.inner.poll_recv(cx)
     }
+
+    /// Returns the bounds of the stream based on the underlying receiver.
+    ///
+    /// For open channels, it returns `(receiver.len(), None)`.
+    ///
+    /// For closed channels, it returns `(receiver.len(), Some(used_capacity))`
+    /// where `used_capacity` is calculated as `receiver.max_capacity() -
+    /// receiver.capacity()`. This accounts for any [`Permit`] that is still
+    /// able to send a message.
+    ///
+    /// [`Permit`]: struct@tokio::sync::mpsc::Permit
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let upper = if self.inner.is_closed() {
+            let used_capacity = self.inner.max_capacity() - self.inner.capacity();
+            Some(used_capacity)
+        } else {
+            None
+        };
+
+        (self.inner.len(), upper)
+    }
```

Comment:
Downstream developer may not send anything through the `Permit`, and then drop it. We can't assume every `Permit` always send a value.

Reply:
I understand and still think the logic is correct. We use inner.len(), which represents values actually waiting in the channel as a lower bound for values count. According to bounded `Receiver`'s docs, `max_capacity()` always returns value specified when declaring the channel and `capacity()` returns `max_capacity()` reduced by number of values waiting in the channel and number of permits that are still able to send. So this method in that case returns values waiting in the channel as a lower bound. For upper bound we have values waiting in the channel plus existing `Permits` that still can send.

Reply:
I expect the stream always yield the exact same (as the upper bound) number of item.

Reply:
Emm, I think you are right, I confused this with another topics. I will close this coversation.